daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -317,53 +430,50 @@ private[lf] object SBuiltin {
         case SDate(date) => date.toString
         case SBigNumeric(x) => Numeric.toString(x)
         case SContractId(_) | SNumeric(_) => crash("litToText: literal not supported")
-      })
+        case otherwise =>
+          throw SErrorCrash(
+            s"${getClass.getSimpleName}: type mismatch of argument 0: expect SNumeric but got $otherwise"
+          )
+      }
+      SText(string)
     }
   }
 
   final case object SBToTextContractId extends SBuiltin(1) {
     override private[speedy] final def execute(
         args: util.ArrayList[SValue],
         machine: Machine,
-    ): Unit = {
-      args.get(0) match {
-        case SContractId(cid) =>
-          machine.ledgerMode match {
-            case _: OnLedger =>
-              machine.returnValue = SValue.SValue.None
-            case _: OffLedger.type =>
-              machine.returnValue = SOptional(Some(SText(cid.coid)))
-          }
-        case _ => crash(s"type mismatch toTextContractId: $args")
+    ): Unit =
+      machine.ledgerMode match {
+        case OffLedger =>
+          machine.returnValue = SOptional(Some(SText(getSContractId(args, 0).coid)))
+        case _ =>
+          machine.returnValue = SValue.SValue.None
```

Comment:
This is a change in semantic, you now won’t crash if this is used incorrectly in on-ledger mode. I don’t really see an issue with this but was this intentional?

Reply:
It was intentional. i will revert it as it is not really in the general scope of the PR.

Reply:
fine to keep it as well, just wanted to doublecheck